### PR TITLE
PowerVS: Introduce retry on retryable errors on API call failure

### PR DESCRIPTION
**Pull request by dharaneeshvrd**
**What this PR does / why we need it**:
Fixes https://issues.redhat.com/browse/MULTIARCH-2943

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: cmd/infra/powervs/create.go [hunk 1]
```diff
@@ -1159,13 +1163,28 @@ func (infra *Infra) setupPowerVSDHCP(ctx context.Context, options *CreateInfraOp
 	return nil
 }
 
+// isTimeoutError validates err is timeout error or not, if yes increase the retryCount, if not return the err
+// retry limit error is thrown when retryCount reaches the limit
+func isTimeoutError(err error, currentRetry *int) error {
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1168]: I'm wondering if we can make this generic so that can be used with any API's error..

File: cmd/infra/powervs/create.go [hunk 2]
```diff
@@ -1159,13 +1163,28 @@ func (infra *Infra) setupPowerVSDHCP(ctx context.Context, options *CreateInfraOp
...
+	if strings.Contains(err.Error(), "status 522") || strings.Contains(err.Error(), "status 524") {
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1169]: Is there any other way we can avoid the strings match?

File: cmd/infra/powervs/create.go [hunk 3]
```diff
@@ -77,6 +77,9 @@ const (
 	nodePoolManagementCreds         = "node-management-creds"
 	ingressOperatorCreds            = "ingress-creds"
 	storageOperatorCreds            = "storage-creds"
+
+	// Retry limit on timeout error
+	timeoutErrorRetryLimit = 5
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:82]: are we planning to have this for other api or at moment only for the dhcpserver get call?

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cmd/infra/powervs/create.go (hunk 3 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: Currently used this retry on `power-go-client/clients` GET call on DHCP and cloud connection client while monitoring, since faced these flaky issues only with these clients.

File: cmd/infra/powervs/create.go [hunk 4]
```diff
@@ -77,8 +77,13 @@ const (
 	nodePoolManagementCreds         = "node-management-creds"
 	ingressOperatorCreds            = "ingress-creds"
 	storageOperatorCreds            = "storage-creds"
+
+	// Retry limit
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:81]: Add more information about which retry we are talking about, can be a generic comment.

File: cmd/infra/powervs/create.go [hunk 5]
```diff
@@ -77,8 +77,13 @@ const (
...
+	retryLimit = 5
 )
 
+var timeoutErrorKeywords = []string{"status 522", "status 524"}
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:85]: I see another variable in the line no.33, wondering if we can merge both of these create a new var group.

File: cmd/infra/powervs/create.go [hunk 6]
```diff
@@ -1292,10 +1321,13 @@ func (infra *Infra) isCloudConnectionReady(ctx context.Context, options *CreateI
...
 	}
 
+	var retryCount int
 	f := func() (bool, error) {
 		cloudConn, err = client.Get(infra.CloudConnectionID)
 		if err != nil {
-			return false, err
+			if err = isRetryableError(err, timeoutErrorKeywords, &retryCount); err != nil {
+				return false, err
+			}
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1330]: If its hitting 522 or 524 then it should return function `false, nil` and go for next iteration right instead of proceeding further?

File: cmd/infra/powervs/create.go [hunk 7]
```diff
@@ -1159,13 +1165,35 @@ func (infra *Infra) setupPowerVSDHCP(ctx context.Context, options *CreateInfraOp
...
+			break
+		}
+	}
+	if retry {
+		if *retryCount > retryLimit {
+			return fmt.Errorf("retry limit exceeded on error: %w", err)
+		}
+		*retryCount += 1
+		return nil
+	}
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1184]: I feel pooling method does a same task, can we have ignore error only part of this flow and let polling method to let polling method itself to deal with this based on the timeout.

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cmd/infra/powervs/create.go (hunk 4 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: Added.

File: cmd/infra/powervs/create.go (hunk 5 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: Grouped all the vars

File: cmd/infra/powervs/create.go (hunk 6 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: yeah that's the intention, I have refactored to look more clear, ptal.

File: cmd/infra/powervs/create.go (hunk 7 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: I thought of limiting the retryable errors too with some limit. If we skip in case of retryable error and depend on the timeout of polling method, it may retry on the whole duration until the timeout occurs. In case of DHCP monitoring we have timeout of 30mins.

Comment (mkumatag) [reply cmd/infra/powervs/create.go]: if it is not that flaky then we are good, if its fails forever then we have an issues. Have you come across a scenario it failed always with 5xx error in that window.?

File: cmd/infra/powervs/create.go [hunk 8]
```diff
@@ -1159,13 +1165,40 @@ func (infra *Infra) setupPowerVSDHCP(ctx context.Context, options *CreateInfraOp
 	return nil
 }
 
+// isRetryableError validates err contains possible retryable error keywords like timeoutErrorKeywords, if yes, the action can be retried and increase the retryCount, if not return the error.
+// if retryCount exceeds retryLimit, a retry limit exceeded error will be thrown.
+func isRetryableError(err error, retryableErrorKeywords []string, retryCount *int) (bool, error) {
```
Comment (mkumatag) [inline cmd/infra/powervs/create.go:1170]: do we need bool here? I see always we returned `false, err` or `true, nil`, I think we can remove bool return.

File: cmd/infra/powervs/create.go (hunk 7 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: Not really, I have n't seen its consistently repeating.

Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: Ok, then maybe will remove the retryLimit logic and in future if we want to limit for some reason based on some scenario, we can maintain this count on calling func itself I guess.

Comment (dharaneeshvrd) [reply cmd/infra/powervs/create.go]: refactored the code, ptal!

File: cmd/infra/powervs/destroy.go [hunk 9]
```diff
@@ -416,11 +419,14 @@ func destroyPowerVsCloudInstance(ctx context.Context, options *DestroyInfraOptio
 
 // monitorPowerVsJob monitoring the submitted deletion job
 func monitorPowerVsJob(id string, client *instance.IBMPIJobClient, infraID string, timeout time.Duration) error {
-
 	f := func() (bool, error) {
 		job, err := client.Get(id)
 		if err != nil {
-			return false, err
+			if err := isNotRetryableError(err, timeoutErrorKeywords); err != nil {
```
Comment (mkumatag) [inline cmd/infra/powervs/destroy.go:425]: here I see you are declaring the local `err` but not in other place, can this be consistent everywhere?

**Review by mkumatag — CHANGES_REQUESTED** [review]
small nit, otherwise lgtm

File: cmd/infra/powervs/destroy.go (hunk 9 above)
Comment (dharaneeshvrd) [reply cmd/infra/powervs/destroy.go:425]: Modified, Ptal now!

File: cmd/infra/powervs/create.go [hunk 10]
```diff
@@ -105,16 +98,25 @@ type TimeDuration struct {
 	time.Duration
 }
 
-var unsupportedPowerVSZones = []string{"wdc06"}
+var (
+	cloudApiKey             string
```
Comment (enxebre) [inline cmd/infra/powervs/create.go:102]: are these changes related to this PR at all?
